### PR TITLE
fix: server request event retention during async handling

### LINKED ISSUE #4417: [🐞] `this` is undefined in server$ function when called consequentially.
### Which component is affected?

Qwik City (routing)

### Describe the bug

When two or more `$server` functions called one after each other the second called function would have `this === undefined`

This doesn't happen when they called concurrently in `Promise.all`

```ts
const serverFunctionA = server$(async function () {
  console.log('serverFunctionA', this);
  return true;
});

const serverFunctionB = server$(async function () {
  console.log('serverFunctionB', this); // <--- here `this` would be undefined
  return true;
});

export default component$(() => {
  const resource = useResource$(async () => {
    const resultA = await serverFunctionA();
    const resultB = await serverFunctionB();

    return { resultA, resultB };
  });

  return (
    <Resource
      value={resource}
      onResolved={({ resultA, resultB }) => (
        <>
          {resultA} {resultB}
        </>
      )}
    />
  );
});
```

### Reproduction

https://stackblitz.com/edit/qwik-starter-jdcnqd?file=src%2Froutes%2Findex.tsx

### Steps to reproduce

_No response_

### System Info

```shell
System:
    OS: macOS 13.3
    CPU: (10) arm64 Apple M1 Pro
    Memory: 75.86 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 16.17.0 - ~/.nvm/versions/node/v16.17.0/bin/node
    Yarn: 3.5.1 - ~/.nvm/versions/node/v16.17.0/bin/yarn
    npm: 8.15.0 - ~/.nvm/versions/node/v16.17.0/bin/npm
    Watchman: 2023.02.27.00 - /opt/homebrew/bin/watchman
  Browsers:
    Chrome: 113.0.5672.126
    Firefox: 95.0.2
    Safari: 16.4
  npmPackages:
    @builder.io/qwik: 1.1.4 => 1.1.4 
    @builder.io/qwik-city: ~1.1.4 => 1.1.4 
    @builder.io/qwik-react: ^0.5.0 => 0.5.0 
    undici: 5.21.0 => 5.21.0 
    vite: 4.2.1 => 4.2.1
```

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by wmertens**
fixes #4417 

This passes one test already, but it needs to use a global to communicate between the qwik-city middleware and the qwik-city server functions, so it's not ideal.

### CODE REVIEW

File: packages/qwik-city/middleware/request-handler/user-response.ts [hunk 1]
```diff
@@ -13,6 +13,16 @@ export interface QwikCityRun<T> {
   completion: Promise<unknown>;
 }
 
+let asyncStore: import('node:async_hooks').AsyncLocalStorage<RequestEventInternal> | undefined
+import('node:async_hooks').then(module => {
```
Comment (PatrickJS) [inline packages/qwik-city/middleware/request-handler/user-response.ts:17]: ideally we use top-level await but ran into type issues

File: packages/qwik-city/middleware/request-handler/user-response.ts [hunk 2]
```diff
@@ -37,7 +47,7 @@ export function runQwikCity<T>(
   return {
     response: responsePromise,
     requestEv,
-    completion: runNext(requestEv, resolve!),
+    completion: asyncStore ? console.error('eyeyya')||asyncStore.run(requestEv, runNext,requestEv, resolve!): runNext(requestEv, resolve!),
```
Comment (mhevery) [inline packages/qwik-city/middleware/request-handler/user-response.ts:50]: nit: left behind a console.

File: packages/qwik-city/runtime/src/server-functions.ts [hunk 3]
```diff
@@ -287,7 +287,9 @@ export const serverQrl = <T extends ServerFunction>(qrl: QRL<T>): ServerQRL<T> =
           : undefined;
       if (isServer) {
         // Running during SSR, we can call the function directly
-        const requestEvent = [useQwikCityEnv()?.ev, this, _getContextEvent()].find(
+        const foo = [(globalThis as any).asyncStore?.getStore(), useQwikCityEnv()?.ev, this, _getContextEvent()]
+        console.trace('asdfasdfasdfasdfasdf', foo.map(Boolean))
```
Comment (mhevery) [inline packages/qwik-city/runtime/src/server-functions.ts:291]: nit: console

File: starters/apps/qwikcity-test/src/routes/(common)/server-func/index.tsx [hunk 4]
```diff
@@ -78,6 +79,29 @@ export default component$(() => {
...
+  return this.method;
+});
+const serverFunctionB = server$(async function b() {
+  return this.method;
+});
+
+export const MultipleServerFunctionsInvokedInTask = component$(() => {
+  const methods = useStore<{ a: string; b: string }>({ a: "", b: "" });
+  useTask$(async () => {
+    methods.a = await serverFunctionA();
```
Comment (mhevery) [inline starters/apps/qwikcity-test/src/routes/(common)/server-func/index.tsx:97]: to make sure you are running async, add `await delay(1)`

**Review by mhevery — COMMENTED** [review]
Yes, this is the approach which we have discussed.

**Comment by PatrickJS** [comment]
I'm updating the tests

File: packages/qwik-city/middleware/request-handler/request-event.ts [hunk 5]
```diff
@@ -73,7 +73,9 @@ export function createRequestEvent(
 
     while (routeModuleIndex < requestHandlers.length) {
       const moduleRequestHandler = requestHandlers[routeModuleIndex];
-      const result = moduleRequestHandler(requestEv);
+      const result = (globalThis as any).asyncStore
+        ? (globalThis as any).asyncStore.run(requestEv, moduleRequestHandler, requestEv)
```
Comment (mhevery) [inline packages/qwik-city/middleware/request-handler/request-event.ts:77]:
```suggestion
        ? (globalThis as QGlobal).asyncStore.run(requestEv, moduleRequestHandler, requestEv)
```

```
interface QGlobal extends Global {
  asyncStore?: AsyncLocalStorage;
}
```

File: packages/qwik-city/middleware/request-handler/user-response.ts [hunk 6]
```diff
@@ -13,6 +13,21 @@ export interface QwikCityRun<T> {
   completion: Promise<unknown>;
 }
 
+let asyncStore: import('node:async_hooks').AsyncLocalStorage<RequestEventInternal> | undefined;
+import('node:async_hooks')
+  .then((module) => {
+    const AsyncLocalStorage = module.AsyncLocalStorage;
+    asyncStore = new AsyncLocalStorage<RequestEventInternal>();
+    // TODO add type
+    (globalThis as any).asyncStore = asyncStore;
```
Comment (mhevery) [inline packages/qwik-city/middleware/request-handler/user-response.ts:22]:
```suggestion
    (globalThis as QGlobal).asyncStore = asyncStore;
```

File: packages/qwik-city/runtime/src/server-functions.ts [hunk 7]
```diff
@@ -287,7 +287,13 @@ export const serverQrl = <T extends ServerFunction>(qrl: QRL<T>): ServerQRL<T> =
           : undefined;
       if (isServer) {
         // Running during SSR, we can call the function directly
-        const requestEvent = [useQwikCityEnv()?.ev, this, _getContextEvent()].find(
+        const contexts = [
+          (globalThis as any).asyncStore?.getStore(),
```
Comment (mhevery) [inline packages/qwik-city/runtime/src/server-functions.ts:291]:
```suggestion
          (globalThis as QGlobal).asyncStore?.getStore(),
```

**Review by mhevery — APPROVED** [review]
Minor comments, LGTM otherwise.

File: packages/qwik-city/middleware/request-handler/request-event.ts [hunk 8]
```diff
@@ -27,6 +27,12 @@ import type { QwikManifest, ResolvedManifest } from '@builder.io/qwik/optimizer'
 import { IsQData, QDATA_JSON, QDATA_JSON_LEN } from './user-response';
 import { isPromise } from './../../runtime/src/utils';
 
+// TODO: create single QGlobal type
+type AsyncStore = import('node:async_hooks').AsyncLocalStorage<RequestEventInternal>;
```
Comment (mhevery) [inline packages/qwik-city/middleware/request-handler/request-event.ts:31]: Why are you doing a dynamic import here? Why not just regular static import? See https://developers.cloudflare.com/workers/runtime-apis/nodejs/asynclocalstorage/

Comment (wmertens) [reply packages/qwik-city/middleware/request-handler/request-event.ts:31]: Because if it's running somewhere without ALS, it can behave as before

**Pull request merged**